styx-api-service/src/main/java/com/spotify/styx/api/StatusResource.java

```diff
@@ -146,25 +154,36 @@ private static String arg(String name, RequestContext rc) {
 
   private Map<WorkflowInstance, RunState> getActiveStates(String componentsStr) throws IOException {
     final List<String> components = Arrays.asList(componentsStr.split(","));
-    final ImmutableMap.Builder<WorkflowInstance, RunState> mapBuilder = ImmutableMap.builder();
-
-    final List<Optional<IOException>> exceptions = components.parallelStream().map( componentId -> {
-      Optional<IOException> exception = Optional.empty();
-      try{
-        final Map<WorkflowInstance, RunState> stateMap = storage.readActiveStates(componentId);
-        for (WorkflowInstance instance : stateMap.keySet()) {
-          mapBuilder.put(instance, stateMap.get(instance));
-        }
-      } catch (IOException e) {
-        exception = Optional.of(e);
-      }
-      return exception;
-      }).filter(Optional::isPresent).collect(toList());
+    final ConcurrentHashMap<WorkflowInstance, RunState> concurrentHashMap = new ConcurrentHashMap<>();
+
+    final List<Optional<IOException>> exceptions =
+        components.stream()
+            .map(
+                componentId ->
+                    forkJoinPool.submit(
+                        () -> {
+                          Optional<IOException> exception = Optional.empty();
+                          try {
+                            final Map<WorkflowInstance, RunState> stateMap =
+                                storage.readActiveStates(componentId);
+                            for (WorkflowInstance instance : stateMap.keySet()) {
+                              concurrentHashMap.put(instance, stateMap.get(instance));
+                            }
+                          } catch (IOException e) {
+                            exception = Optional.of(e);
+                          }
+                          return exception;
+                        }))
+            .collect(toList())
+            .stream()
+            .map(ForkJoinTask::join)
+            .filter(Optional::isPresent)
+            .collect(toList());
```

Comment:
We should be able to rewrite this logic to collect a map instead of a list of exceptions and then return the first one if not empty. So we can avoid creating a concurrent map.

Reply:
e.g. https://stackoverflow.com/questions/42490314/convert-list-of-maps-to-single-map-via-streams

To handle checked exception, we could use `Try` from javaslang, e.g. https://github.com/spotify/styx/blob/master/styx-scheduler-service/src/main/java/com/spotify/styx/state/PersistentStateManager.java#L122

Reply:
Alternatively we could turn this thing around. Instead of building a map, we could build a list of exceptions and later on check whether the list is empty or not.